src/ILOPoolSale.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+pragma solidity =0.7.6;
+pragma abicoder v2;
+
+import { ILOPoolBase, IILOPoolBase } from './base/ILOPoolBase.sol';
+import { ILOWhitelist } from './base/ILOWhitelist.sol';
+import { IILOPoolSale } from './interfaces/IILOPoolSale.sol';
+import { IILOManager } from './interfaces/IILOManager.sol';
+import { TransferHelper } from './libraries/TransferHelper.sol';
+import { PoolAddress } from './libraries/PoolAddress.sol';
+import { FullMath } from '@uniswap/v3-core/contracts/libraries/FullMath.sol';
+
+contract ILOPoolSale is ILOPoolBase, IILOPoolSale, ILOWhitelist {
+    bool public override CANCELLED;
+    uint64 public override SALE_START;
+    uint64 public override SALE_END;
+    uint256 public override MIN_RAISE;
+    uint256 public override MAX_RAISE;
+    uint256 public override TOTAL_RAISED;
+
+    VestingConfig private _vestingConfig;
+
+    modifier onlyProjectAdmin() override {
+        IILOManager.Project memory _project = IILOManager(MANAGER).project(
+            PROJECT_ID
+        );
+        require(msg.sender == _project.admin, 'UA');
+        _;
+    }
+
+    modifier duringSale() {
+        require(SALE_START <= block.timestamp, 'BS');
+        require(SALE_END > block.timestamp, 'ES');
+        _;
+    }
+
+    modifier whenCancelled() {
+        require(CANCELLED, 'NOT_CANCELLED');
+        _;
+    }
+
+    modifier whenNotCancelled() {
+        require(!CANCELLED, 'CANCELLED');
+        _;
+    }
+
+    function initialize(InitParams calldata params) external override {
+        _initialize(params.poolParams);
+        require(params.poolParams.vestingConfigs.length == 1, 'VC');
+        _vestingConfig = params.poolParams.vestingConfigs[0];
+        (SALE_START, SALE_END, MIN_RAISE, MAX_RAISE) = (
+            params.saleParams.start,
+            params.saleParams.end,
+            params.saleParams.minRaise,
+            params.saleParams.maxRaise
+        );
+    }
+
+    function buy(
+        uint256 raiseAmount,
+        address recipient
+    ) external override duringSale whenNotCancelled returns (uint256 tokenId) {
+        require(raiseAmount > 0, 'ZA');
+        // check if raise amount over capacity
+        uint256 totalRaised = TOTAL_RAISED;
+        require(MAX_RAISE - totalRaised >= raiseAmount, 'MR');
+        totalRaised += raiseAmount;
+        TOTAL_RAISED = totalRaised;
```

Comment:
only increase totalRaised after transfer from

Reply:
increase before transfer_from is safer. 
but better to add the reentrancy guard